### ISSUE TITLE
Memory leak in the latest netty version when executing tests

### ISSUE THREAD

**Issue by doom369** [netty/netty#8127]
I just updated to the latest netty version (4.1.26.Final, boring.ssl  2.0.12.Final) and my test run had crashed due to OutOfMemoryError.
I have a special test set with 200 tests. Every test starts the server and at the end of test stops it. Before tests were executed just fine. With the latest version on 100-th test, my RAM of 32 GBs is out. 

I haven't dig into it yet, just let you know. Probably this is related to recent "finalizer" changes and object cleaner.

```
java version "10.0.1" 2018-04-17
Java(TM) SE Runtime Environment 18.3 (build 10.0.1+10)
Java HotSpot(TM) 64-Bit Server VM 18.3 (build 10.0.1+10, mixed mode)
```

Prev version was:

4.1.25.Final
boring.ssl 2.0.8.Final

**Comment by normanmaurer** [comment]
@doom369 can you show me how you configure your `SslContextBuilder` ?

**Comment by normanmaurer** [comment]
also what version you used before ?

**Comment by doom369** [comment]
@normanmaurer updated ticket with prev version. I build it with this - https://github.com/blynkkk/blynk-server/blob/master/server/core/src/main/java/cc/blynk/server/SslContextHolder.java

**Comment by normanmaurer** [comment]
I wonder if you can try to run the tests with `SslProvider.JDK` and if you can try to run it with `SslProvider.OPENSSL` but not use a key file but use a `KeyManagerFactory` ?

**Comment by doom369** [comment]
@normanmaurer tried `SslProvider.JDK`, issue still present.

**Comment by doom369** [comment]
`SslProvider.JDK` doesn't help.

**Comment by normanmaurer** [comment]
interesting... Would it be possible get a heap dump ?

**Comment by normanmaurer** [comment]
Also I wonder if you could try to just revert `5b1fe611a637c362a60b391079fff73b1a4ef912` and see if you still see the issue. At the moment I am not sure what the cause is really :(

**Comment by doom369** [comment]
I thinks this should give you some clues:

![screenshot from 2018-07-11 17-24-37](https://user-images.githubusercontent.com/1536494/42578629-62ba6214-852f-11e8-87b7-8fe3dbfee4d1.png)

**Comment by normanmaurer** [comment]
hmm... can you tell me what holds the reference to the `byte[]` and also how many `EventLoop` instances you have ? It seems you may have 500k+ of these ?

**Comment by doom369** [comment]
@normanmaurer ok, narrowed down the issue to

```
        this.asyncHttpClient = new DefaultAsyncHttpClient(new DefaultAsyncHttpClientConfig.Builder()
                .setUserAgent(null)
                .setKeepAlive(true)
                .setUseNativeTransport(true)
                .setUseOpenSsl(true)
                .build()
        );
```

After setting `.setUseNativeTransport(false)` all is ok. Probably issue is because `AsyncHttpClient` uses prev. netty version. Sorry for false alarm.

**Comment by normanmaurer** [comment]
@doom369 ok... I wonder if the async http client not tears down the event loop correctly and so its possible the native memory is not released.

**Comment by doom369** [comment]
@normanmaurer could it be because of changed native bindings to epoll transport?

**Comment by normanmaurer** [comment]
@doom369 I don't think so... The only reason I can think of is because of this change:

https://github.com/netty/netty/commit/05e5ab1ecb98963604d686c1f59b2196cf73e244#diff-824cfcec3522343772416dd064658a83

As now we reserve the native memory on `EventLoop` creation directly while before it was only reserved on first usage via `EpollDatagramChannel`. This memory will be released once you call `EpollEventLoopGroup.shutdownGracefully()` tho, so I suspect `AsyncHttpClient ` not tears down its EventLoops correctly.

**Comment by doom369** [comment]
Ok. Thanks I'll move this to AsynClient Report.

**Comment by normanmaurer** [comment]
@doom369 having `500k+` instances does not look right to me, or I am reading it wrong in the screenshot ?

**Comment by doom369** [comment]
I don't know what `IovArray` is, so I can't tell :). The thing is - I never closed explicitly the `asyncHttpClient` between my tests run and seems like that's a reason for a leak. I thought that when it GCed it should release all things it holds, but seems like this is not true anymore. Adding `asyncHttpClient.close();` fixes the issue.

**Comment by normanmaurer** [comment]
@doom369 it was also not true before it just leaked less as it only leaked IovArray before and now it also leaks NativeDatagramPacket ;)

**Comment by normanmaurer** [comment]
@doom369 thanks for the help to debug this :)

**Comment by doom369** [comment]
@normanmaurer thank you too! Will do digging before posting next time :).

**Comment by slandelle** [comment]
> Probably issue is because AsyncHttpClient uses prev. netty version.

I released AHC 2.5.1 w/ Netty 4.1.26 earlier today :)
